### PR TITLE
verbs: Bug fixes

### PR DESCRIPTION
**Pull request by a-ilango**
Contents:
- [Util] Add macros for accessing fields in sockaddr_in* structs
- Return all valid verbs interfaces on a system when user doesn't specify a node, or any addressing info and FI_SOURCE is passed.
- Fix incorrect reporting cq_size.

Update: Dropped patch 1 and 3 from this PR.

### CODE REVIEW

**Comment by a-ilango** [comment]
@evgeny-leksikov Let me know if this would work fine with verbs/RDM.

**Comment by evgeny-leksikov** [comment]
@a-ilango I see regression after 8a21bd8 of fi_av_test test for release build (debug build works) on finalization:

>...
> Summary: all tests passed
> *** glibc detected *** fi_av_test: corrupted double-linked list: 0x000000000203aa40 ***

( at ../prov/verbs/src/verbs_domain.c:344)

Also, it breaks Intel MPI initialization in fi_getinfo at ../src/fabric.c:561 (*info, tail, cur are NULL)

**Comment by a-ilango** [comment]
@evgeny-leksikov Thank you for testing. Interestingly, I don't see the glibc corruption issue when I run fi_av_test on a non-debug build. Valgrind doesn't show any related issues either. How does Intel MPI initialization call fi_getinfo? What args are getting passed?

**Comment by shefty** [comment]
Where do we stand on this PR?  Is it ready, or does it still break MPI?

**Comment by a-ilango** [comment]
It's not ready yet. The PR breaks MPI initialization and I'm still investigating it.

**Comment by shefty** [comment]
Can you separate the CQ fix into its own PR, so that can be merged?

File: include/fi_util.h [hunk 1]
```diff
@@ -87,6 +87,9 @@
 	OFI_Q_READERR(prov, log, eq, "eq", fi_eq_readerr, 	\
 			fi_eq_strerror, ret, err_entry)
 
+#define ofi_sin_addr(addr) (((struct sockaddr_in *)(addr))->sin_addr)
+#define ofi_sin6_addr(addr) (((struct sockaddr_in6 *)(addr))->sin6_addr)
```
Comment (a-ilango) [inline include/fi_util.h]: I'm using it in patch # 2. They can be used in other places as well.

**Comment by a-ilango** [comment]
I'll create a separate PR for the CQ fix and drop that from here.

File: include/fi_util.h (hunk 1 above)
Comment (shefty) [inline include/fi_util.h]: Okay - I completely missed where those were used.  If you want to include this patch in the other PR, that's fine too

**Comment by shefty** [comment]
What would happen if this fix didn't make it into 1.4 (at least rc1)?  Is this a critical issue?

**Comment by a-ilango** [comment]
This is not a critical fix but it's good to have this in 1.4 release. Also fi_pingpong test wouldn't pass without this fix. Issue #2307 cannot be closed without this.

When the app doesn't pass any addressing parameters, verbs provider either returns a localhost or wildcard address in the info->src_addr. This fix modifies verbs provider to return the interface address (if configured).

**Comment by shefty** [comment]
> Also fi_pingpong test wouldn't pass without this fix. Issue #2307
> <https://github.com/ofiwg/libfabric/issues/2307>  cannot be closed
> without this.

Okay - getting this in for 1.4 is fine, but I'll start preparing rc1 now.  My understanding of #2307 is that fi_pingpong is broken.  The provider should be able to return a wildcard address, especially for passive interfaces.  Fi_pingpong should eliminate its out of band address exchange.

**Comment by evgeny-leksikov** [comment]
@a-ilango I do not see any regressions from quick testing. 👍

**Comment by evgeny-leksikov** [comment]
BTW, should we use this way to define IPoIB address? It looks a bit duplicated with [this code](https://github.com/ofiwg/libfabric/blob/0ce097fc2a61a742f7d4a8ab19acc9f94aabc585/prov/verbs/src/ep_rdm/verbs_utils.c#L188) but more intelligent. And I think this change should also handle FI_VERBS_IFACE parameter.

**Comment by shefty** [comment]
Was there an update to this patch between it being reported as not working versus working now?  I'm confused.

**Comment by evgeny-leksikov** [comment]
@shefty yes, @a-ilango updated this PR at lease twice. We discussed the progress on related issues offline.

**Comment by a-ilango** [comment]
>BTW, should we use this way to define IPoIB address? It looks a bit duplicated with [this code](https://github.com/ofiwg/libfabric/blob/0ce097fc2a61a742f7d4a8ab19acc9f94aabc585/prov/verbs/src/ep_rdm/verbs_utils.c#L188) but more intelligent. And I think this change should also handle FI_VERBS_IFACE parameter.

@evgeny-leksikov Thank you for testing and reporting back. Yes, there is some code duplication and I think the FI_VERBS_IFACE could make use of this. The env variable helps the user to run on a particular interface on a node, right? fi_ibv_getinfo can be modified to return only the fi_info corresponding to the interface in the env variable. I think fi_ibv_getinfo would need some refactoring as well which could be done after 1.4

**Pull request merged**